### ISSUE TITLE
Workload Listener does not Have SDS Config when SDS and CNI are turned on Both

### ISSUE THREAD

**Issue by stiller-leser** [istio/istio#15701]
(NOTE: This is used to report product bugs:
  To report a security vulnerability, please visit <https://istio.io/about/security-vulnerabilities/>
  To ask questions about how to use Istio, please visit <https://discuss.istio.io>
)

**Bug description**

I am trying to set up my Istio cluster with MUTUAL_TLS and SDS activated. All requests to my sample nginx pod over the ingressgateway result in a 503. Mutual TLS configuration seems to be ok, according to 

```
istioctl authn tls-check istio-ingressgateway-5959c8c7d8-hzf66 nginx1.default.svc.cluster.local -n istio-system
HOST:PORT                               STATUS     SERVER     CLIENT     AUTHN POLICY     DESTINATION RULE
nginx1.default.svc.cluster.local:80     OK         mTLS       mTLS       default/         default/istio-system
```

Nginx pod can be correctly reached from within the cluster as long as the ingressgateway is not used.

Logs from ingressgateway show:

```
[2019-07-09 09:07:24.907][29][debug][pool] [external/envoy/source/common/http/http1/conn_pool.cc:88] creating a new connection
[2019-07-09 09:07:24.907][29][debug][client] [external/envoy/source/common/http/codec_client.cc:26] [C4759] connecting
[2019-07-09 09:07:24.907][29][debug][connection] [external/envoy/source/common/network/connection_impl.cc:702] [C4759] connecting to 100.200.1.59:80
[2019-07-09 09:07:24.907][29][debug][connection] [external/envoy/source/common/network/connection_impl.cc:711] [C4759] connection in progress
[2019-07-09 09:07:24.907][29][debug][pool] [external/envoy/source/common/http/conn_pool_base.cc:20] queueing request due to no available connections
[2019-07-09 09:07:24.907][29][debug][connection] [external/envoy/source/common/network/connection_impl.cc:550] [C4759] connected
[2019-07-09 09:07:24.907][29][debug][connection] [external/envoy/source/extensions/transport_sockets/tls/ssl_socket.cc:168] [C4759] handshake error: 2
[2019-07-09 09:07:24.907][29][debug][connection] [external/envoy/source/extensions/transport_sockets/tls/ssl_socket.cc:168] [C4759] handshake error: 1
[2019-07-09 09:07:24.907][29][debug][connection] [external/envoy/source/extensions/transport_sockets/tls/ssl_socket.cc:201] [C4759] TLS error: 268435703:SSL routines:OPENSSL_internal:WRONG_VERSION_NUMBER
[2019-07-09 09:07:24.907][29][debug][connection] [external/envoy/source/common/network/connection_impl.cc:188] [C4759] closing socket: 0
[2019-07-09 09:07:24.907][29][debug][client] [external/envoy/source/common/http/codec_client.cc:82] [C4759] disconnect. resetting 0 pending requests
[2019-07-09 09:07:24.907][29][debug][pool] [external/envoy/source/common/http/http1/conn_pool.cc:129] [C4759] client disconnected, failure reason: TLS error: 268435703:SSL routines:OPENSSL_internal:WRONG_VERSION_NUMBER
[2019-07-09 09:07:24.907][29][debug][pool] [external/envoy/source/common/http/http1/conn_pool.cc:164] [C4759] purge pending, failure reason: TLS error: 268435703:SSL routines:OPENSSL_internal:WRONG_VERSION_NUMBER
[2019-07-09 09:07:24.907][29][debug][router] [external/envoy/source/common/router/router.cc:671] [C4753][S3527573287149425977] upstream reset: reset reason connection failure
[2019-07-09 09:07:24.907][29][debug][http] [external/envoy/source/common/http/conn_manager_impl.cc:1137] [C4753][S3527573287149425977] Sending local reply with details upstream_reset_before_response_started{connection failure,TLS error: 268435703:SSL routines:OPENSSL_internal:WRONG_VERSION_NUMBER}
```

Additionally, since I am using SDS, there won't be any certificates in ingressgateway-certs (see https://istio.io/docs/tasks/security/auth-sds/#verifying-no-secret-volume-mounted-file-is-generated) as it is described in https://istio.io/docs/tasks/traffic-management/ingress/secure-ingress-mount/

If I deactivate `MUTUAL_TLS` in the default destination rule, things start working as intented. 

`istioctl proxy-config endpoint nginx1` shows port 443 as open. `istioctl proxy-config route nginx1` shows routes on port 80 and inbound on `inbound|80|http|nginx1.default.svc.cluster.local` but nothing on 443 (which is okay, since I haven't configured anything).

This might be a configuration issue, however I am having difficulties locating the issue as all of the troubleshooting guides produce the expected positive result. Happy to provide more information if required.

**Affected product area (please put an X in all that apply)**

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[ x ] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

**Expected behavior**

I should reach the nginx pod without issues (code 200)

**Steps to reproduce the bug**

- Install Istio in Version 1.2.0 with the following config:

```
# Istio configuration with the goal of using DynaTrace for monitoring instead of the OSS derivates

global:
  # WILL FAIL IF SDS IS ENABLED!
  # controlPlaneSecurityEnabled disable. Will result in delays starting the pods while secrets are
  # propagated, not recommended for tests.
  controlPlaneSecurityEnabled: false

  # Default mtls policy. If true, mtls between services will be enabled by default.
  mtls:
    # Default setting for service-to-service mtls. Can be set explicitly using
    # destination rules or service annotations.
    enabled: true

  # SDS enabled. IF set to true, mTLS certificates for the sidecars will be
  # distributed through the SecretDiscoveryService instead of using K8S secrets to mount the certificates.
  sds:
    enabled: true
    udsPath: "unix:/var/run/sds/uds_path"
    useNormalJwt: true

  imagePullPolicy: Always

#
# sidecar-injector webhook configuration
#
sidecarInjectorWebhook:
  enabled: true
  rewriteAppHTTPProbe: true
  enableNamespacesByDefault: true

#
# Istio CNI plugin enabled
#   This must be enabled to use the CNI plugin in Istio.  The CNI plugin is installed separately.
#   If true, the privileged initContainer istio-init is not needed to perform the traffic redirect
#   settings for the istio-proxy.
#
istio_cni:
  enabled: true

#
# nodeagent configuration
#
nodeagent:
  enabled: true
  image: node-agent-k8s
  env:
    CA_PROVIDER: "Citadel"
    CA_ADDR: "istio-citadel:8060"
    VALID_TOKEN: true

#
# addon certmanager configuration
# requires:
#
certmanager:
  enabled: true

#
# security configuration
#
security:
  enabled: true
  createMeshPolicy: true

#
# Gateways Configuration
#
gateways:
  istio-ingressgateway:
    resources:
      requests:
        cpu: 10m
        memory: 40Mi
    sds:
      enabled: true
    ports:
      ## You can add custom gateway ports
    - port: 80
      targetPort: 80
      name: http2
    - port: 443
      name: https
  istio-egressgateway:
    enabled: true
    resources:
      requests:
        cpu: 10m
        memory: 40Mi
    ports:
      ## You can add custom gateway ports
    - port: 80
      targetPort: 80
      name: http2
    - port: 443
      name: https
```

- Deploy the following:

```
apiVersion: networking.istio.io/v1alpha3
kind: Gateway
metadata:
  name: default-gateway
  namespace: istio-system
spec:
  selector:
    istio: ingressgateway
  servers:
  - port:
      name: http
      number: 80
      protocol: HTTP
    hosts:
    - '*'
---
apiVersion: v1
kind: Pod
metadata:
  creationTimestamp: null
  labels:
    run: nginx1
  name: nginx1
spec:
  containers:
  - image: nginx
    name: nginx
    resources: {}
    ports:
    - containerPort: 80
    - containerPort: 443
  dnsPolicy: ClusterFirst
  restartPolicy: Never
status: {}
---
apiVersion: v1
kind: Service
metadata:
  labels:
    run: nginx1
  name: nginx1
spec:
  ports:
  - name: http
    port: 80
    protocol: TCP
    targetPort: 80
  - name: https
    port: 443
    protocol: TCP
    targetPort: 443
  selector:
    run: nginx1
  sessionAffinity: None
  type: ClusterIP
status:
  loadBalancer: {}
---
apiVersion: networking.istio.io/v1alpha3
kind: VirtualService
metadata:
  name: nginx1
spec:
  hosts:
  - "*"
  gateways:
  - istio-system/default-gateway
  http:
  - match:
    - uri:
        prefix: /nginx1
    rewrite:
      uri: "/"
    route:
    - destination:
        host: nginx1.default.svc.cluster.local
```

** EDIT **

The issue also occurs if I remove all references to HTTPS and 443.

**Version (include the output of `istioctl version --remote` and `kubectl version`)**

Istio:

```
client version: 1.2.1
citadel version: 1.2.0
egressgateway version: 1.2.0
galley version: 1.2.0
ingressgateway version: 1.2.0
nodeagent version: 
nodeagent version: 
pilot version: 1.2.0
policy version: 1.2.0
sidecar-injector version: 1.2.0
telemetry version: 1.2.0
```
K8s:
```
Client Version: version.Info{Major:"1", Minor:"14", GitVersion:"v1.14.2", GitCommit:"66049e3b21efe110454d67df4fa62b08ea79a19b", GitTreeState:"clean", BuildDate:"2019-05-16T18:55:03Z", GoVersion:"go1.12.5", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"11", GitVersion:"v1.11.9", GitCommit:"16236ce91790d4c75b79f6ce96841db1c843e7d2", GitTreeState:"clean", BuildDate:"2019-03-25T06:30:48Z", GoVersion:"go1.10.8", Compiler:"gc", Platform:"linux/amd64"}
```

**How was Istio installed?**

```
 helm template istio/istio-1.2.0/install/kubernetes/helm/istio --name istio --namespace istio-system --values istio-values.yaml | kubectl apply -f -
```

**Environment where bug was observed (cloud vendor, OS, etc)**

```
Internal cloud vendor running on CoreOS
```

Additionally, please consider attaching a [cluster state archive](http://istio.io/help/bugs/#generating-a-cluster-state-archive) by attaching
the dump file to this issue.

**Comment by stiller-leser** [comment]
This issue has been resolved by not using `istio-cni`. It pretty much boiled down to the fact that the istio-proxy kept trying to reach the container it sidecarded on port 443. Happy to leave this issue open for investigation. If maintainers wish to close it, please feel free to do so.

**Comment by ekarlso** [comment]
I have reproduce this too using Istio installed by the banzai cloud operator and istio 1.2.2

**Comment by incfly** [comment]
First, we confirmed that when SDS is enabled, Istio takes care of ingress gateway already. Ingress as client is able to get certificate using SDS. @Monkeyanator recently verified that, maybe not on 1.2 though.

Second, based on your description, things won't work when "istio-cni" is enabled, and the error message, I suspect it's SDS Workload Certificate does not work well with Istio CNI mode together. I need to verify that. Might related to how the sidecar injects happen, since SDS relies on uds path is moutned, and istio cni also involves changes in sidecar injector.

If that's the case, it's definitely a bug.

**Comment by incfly** [comment]
I verified that once you enable CNI, ingress can't talk with workload in mTLS using SDS.

-I looked at the nginx (server's) listener config, turns out no TLS settings are there. This explains why no DR with mTLS can talk to Authn.STRICT together.
- The `istioctl auth-check tls` is misleading, because AFAIK, it checks the policy, rather than the Envoy config. I remember @yangminzhu made this point a while ago, we'd better add two more columns in `istioctl auth-check tls` to show the client and server actual envoy config, rather than infer based on policy statically.

Pass to SDS and CNI owner. I saw the nginx pod has `/var/run/sds` host path injected when CNI is turned on, might be something else causing the SDS config is not correctly generated for nginx's listener config.
